### ISSUE TITLE
Concept of project

### ISSUE THREAD

**Issue by stephenwf** [digirati-co-uk/iiif-manifest-editor#144]
**Properties of a project**
* Name
* Filename
* Source
* Export settings
* Vault serialisation (or manifest export?)
* Function to convert Resource X -> Project

**To be held in:**

* React context / hook for project details (part of Shell)
* Local storage persistence
* Future remote persistence (for full projects, not just manifests).

**Comment by tomcrane** [comment]
I like this.

Just a word of caution about ensuring that the Manifest Editor (or rather the shell) doesn't lose the ability to have a simple external interface - people like V&A, Delft etc can do quite complex integrations with the Manifest editor without having to commit to supporting the persistence of **projects** as part of that integration (although they certainly could if they wanted to). They can just provide a Publisher impl but still rely on the ME to manage projects.

It would be fine for the ME to manage its projects as described above **using only internal storage**, when it's [loading IIIF](https://github.com/digirati-co-uk/iiif-manifest-editor/wiki/Loading-IIIF-for-Editing) from external providers and [saving IIIF](https://github.com/digirati-co-uk/iiif-manifest-editor/wiki/Saving-IIIF) to external providers, because the institutional/user notion of project in these cases is external to the Manifest Editor - it's part of CMS workflow, publishing processes or whatever and the ME shouldn't impose its concepts on these external processes. Its out-of-the-box localStorage impl is in this scenario inside the ME boundary. 

I think I've confused myself and everyone by using the preview and permalink interchangeably and have a suggestion to perhaps avoid that confusion...

The preview is a built-in ability to push IIIF resources (not projects) to a special external service, so that any other IIIF consumer can see them, with (initially) a simple list of templated strings in config for opening (you might locally add more templated strings to your config). The user doesn't need to know about any management of that preview URL, or be aware of any secrets associated with being able to update it. The ME just produces URLs, magically. The secret that the ME is using to do this - to update an existing preview if it knows it's already got one for the currently-editing Manifest - never needs to leave the browser, it doesn't need to be portable, the ME never needs to make it available to the user.

This works for preview because it simply doesn't matter whether the ME is overwriting the previous preview or creating a new one; if it doesn't have a preview secret for the Manifest being edited currently it just uses the service to get a new preview - and it errs on the side of creating a new one if there's any doubt about whether it's a different Manifest. There's never any need to be able to update an old preview URL if, say, you open the Manifest from the preview URL in a different browser, or after having cleared local storage, or whatever. That old URL just gets left behind. 

We also have the concept (formalised a bit more in the publisher description above) of being able to plug in different persistence mechanisms - publishers - some of which we might provide, some of which others might implement. We might provide a GitHub integration or bridge; we might provide a client (for both loading and saving) of a really simple REST protocol (https://github.com/digirati-co-uk/iiif-manifest-editor/wiki/REST-Protocol). These shouldn't be strongly coupled to persisting _projects_ as well. V&A could read the spec and just decide to implement the REST protocol on its own, or a REST protocol+, that also provides for persistence for projects externally (no longer in local storage). These are different steps, we need to support both scenarios.

The confusion arises because we were also talking about using the mechanism for preview to provide _an example_ of a persistence mechanism - a Google-doc style URL. So something is happening in the Save menu that is _like_ the preview, but not quite - the UI is more complex because now the user has to decide whether or not they are overwriting the previous URL - and the other unspoken issue of how they might update the Manifest at that published URL later, from somewhere else where the secret access key isn't available. They simply don't need to care about that **for preview** (and we've ditched the intermediary modal), but they do for a more formal concept of publishing.

So - for now I think we should **drop the permalink completely as a Save implementation** - and come back to it later when we can attache an external source of identity to it.
We should put a placeholder action/message in Save - "this will be where you plug in GitHub, REST, your CMS, Digirati super-service, etc" but not actually provide a Save at all. And we just make the Preview URLs not expire.

This gives people a tool to play with so they can understand where we are going, it removes some obvious UI friction around the Save process. People can still export IIIF, and they can share the now-permanent preview URLs - so it's immediately useful.

Implementing a permalink service that integrates with social sign-ins, so that we externally store the map of user identifiers to secret update keys, would be a quite simple additional service that we could come back to and design the _particular save interaction for that Publisher_ as describe above.

> We have reached the limitation of this model

I think the model where the ME boundary is just IIIF in and out is extremely powerful - if the ME is using local storage as the project back end (as defined above). It's also more adaptable to external requirements.

> In order to save to a backend - the list of URLs in local storage - you must first publish or preview. The preview is acting as both the preview, the storage and the backend.

* Avoid this confusion by just having Preview which is deliberately not a proper Publish target (because you have no control over whether the target is updated or created new). It's a completely separate Preview feature.
* Remove the Save option for now but think a bit harder about what the interface/contract is:

ME -> Project -> Storage Adaptor -> [your impl here]

... where we have simple IIIF REST impl, GitHub impl, even filesystem impl...

I think for now we should focus our efforts on the user experience of editing and park any further work on a more complex stateful preview environments that are part of or connected to the app in some way - the current completely external preview is fine for now. We can come back to that. UI friction in the editing process is a much higher priority.

Removing the permalink Save implementation completely at this stage removes some of that friction and with long-lasting previews, will be just as useful for people trying it out.

**Comment by stephenwf** [comment]
> Just a word of caution about ensuring that the Manifest Editor (or rather the shell) doesn't lose the ability to have a simple external interface [ ... ] They can just provide a Publisher impl but still rely on the ME to manage projects.

Yes agree completely. At the moment I have the following interfaces
<details>
<summary></summary>

**Comment by stephenwf** [comment]
> Just a word of caution about ensuring that the Manifest Editor (or rather the shell) doesn't lose the ability to have a simple external interface [ ... ] They can just provide a Publisher impl but still rely on the ME to manage projects.

Yes agree completely. At the moment I have the following interfaces
<details>
<summary>typescript interfaces</summary>

```typescript
export interface ProjectBackend {
  saveInterval: number;
  getAllProjects(): Promise<EditorProject[]>;
  getLastProject(): Promise<string | null>;
  setLastProject(id: string): Promise<void>;
  createProject(project: EditorProject): Promise<void>;
  updateProject(project: EditorProject): Promise<void>;
  deleteProject(id: string): Promise<void>;
}

export interface ProjectStorage<S extends Storage> {
  type: string;
  saveInterval: number;

  // Fetching
  getStorage(id: string): Promise<S | null>;
  updateStorage(id: string, storage: S): Promise<void>;
  deleteStorage(id: string): Promise<void>;

  // Runtime
  getLatestStorage(project: EditorProject): S;
  createVaultInstance(project: EditorProject): [Vault, Reference];
  shouldUpdateWithVault(): boolean;
}
```

</details>

Where the **Backend** could be LocalStorage, and the **Storage** could be V&As  CMS. The publishing options could be a deep link to open a "New V&A article with this manifest embedded". (As publish should be for those things not reversible).

> These shouldn't be strongly coupled to persisting projects as well. V&A could read the spec and just decide to implement the REST protocol on its own, or a REST protocol+, that also provides for persistence for projects externally

Yes definitely. For the initial work, the **Backend** will be almost invisible, only showing you things you've worked on in your browser - feeling like a cache. If the V&A implemented a REST protocol, and configured it as per (https://github.com/digirati-co-uk/iiif-manifest-editor/wiki/REST-Protocol) and added it to the ME configuration, it wouldn't change the format of a project immediately. 

Once a project is published using that new configuration, then `project.publishers[]` gets a new entry. So on a project it's an opportunity to be a "instance", maybe holding the permalink that was published, or label or time etc.
 
> I think for now we should focus our efforts on the user experience of editing and park any further work on a more complex stateful preview environments that are part of or connected to the app in some way 

From a feature point of view, I agree completely. We should park most of the implementations and hardening up the shapes of all of these things. From a code perspective, implementing the concepts above answer a lot of questions and confusions that have come up recently.
